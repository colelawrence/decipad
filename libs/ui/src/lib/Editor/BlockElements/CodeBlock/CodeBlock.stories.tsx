import React from 'react';
import { CodeBlock } from './CodeBlock.component';

export default {
  title: 'Editor/Block Elements/CodeBlock',
};

export const C = () => (
  <CodeBlock
    element={{ children: [{ text: 'hello world ' }] }}
    attributes={{
      'data-slate-node': 'element',
      ref: null,
    }}
  >
    The quick brown fox jumps over the lazy dog
  </CodeBlock>
);

C.storyName = 'CodeBlock';
