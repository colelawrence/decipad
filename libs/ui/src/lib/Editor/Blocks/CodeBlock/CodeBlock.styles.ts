import { Code } from '@chakra-ui/layout';
import { chakra } from '@chakra-ui/system';

export const CodeBlockStyles = chakra(Code, {
  baseStyle: {
    w: '100%',
    p: 5,
    borderRadius: 2,
  },
});
