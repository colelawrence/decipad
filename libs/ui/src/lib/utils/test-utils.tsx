import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';

import { Type } from '@decipad/language';
import { theme, ResultsContextProvider } from '../..';

const results = {
  test: {
    type: Type.build({
      type: 'number',
      unit: {
        unit: 'apples',
      },
    } as any),
    value: [20],
    errors: [],
  },
  noResult: null,
};

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      <ResultsContextProvider value={results}>
        {children}
      </ResultsContextProvider>
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
