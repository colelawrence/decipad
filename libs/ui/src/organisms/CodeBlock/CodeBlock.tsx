import { FC, ReactNode } from 'react';
import { css } from '@emotion/react';
import { Result } from '@decipad/language';
import { code, cssVar, grey250 } from '../../primitives';
import { CodeResult } from '..';

const codeBlockStyles = css(code, {
  border: `1px solid ${grey250.rgb}`,

  borderRadius: '12px',
  overflow: 'hidden',

  margin: '16px 0',
});

const codeStyles = css(code, {
  backgroundColor: cssVar('backgroundColor'),
  padding: '4px 8px',
});

const resultStyles = css({
  backgroundColor: cssVar('highlightColor'),
  display: 'grid',
  padding: '8px',
});

interface CodeBlockProps {
  readonly blockId?: string;
  readonly expandedResult?: Result;
  readonly children?: ReactNode;
}

export const CodeBlock = ({
  children,
  expandedResult,
}: CodeBlockProps): ReturnType<FC> => {
  return (
    <div>
      <section css={codeBlockStyles} spellCheck={false}>
        <pre css={codeStyles}>{children}</pre>
        {expandedResult?.value != null && (
          <output css={resultStyles} contentEditable={false}>
            <CodeResult {...expandedResult} />
          </output>
        )}
      </section>
    </div>
  );
};
