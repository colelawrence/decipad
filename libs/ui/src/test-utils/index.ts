export * from './menu';
export * from './language';
