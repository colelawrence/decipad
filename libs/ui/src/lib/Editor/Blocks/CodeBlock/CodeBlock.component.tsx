import { Box, useColorModeValue } from '@chakra-ui/react';
import React from 'react';
import { RenderElementProps } from 'slate-react';
import { Result } from '../../../Result/Result.component';
import { CodeBlockStyles } from './CodeBlock.styles';

export const CodeBlock = ({
  attributes,
  children,
  element,
}: RenderElementProps): JSX.Element => {
  const bg = useColorModeValue('gray.100', 'gray.700');

  return (
    <Box>
      {element.result && <Result {...(element.result as any)} />}
      <CodeBlockStyles bg={bg} {...attributes}>
        {children}
      </CodeBlockStyles>
    </Box>
  );
};
