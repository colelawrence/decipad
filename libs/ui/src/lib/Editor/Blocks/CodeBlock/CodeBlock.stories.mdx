import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';
import { CodeBlock } from './CodeBlock.component';
import { Type } from '@decipad/language';
import { ResultsContextProvider } from '@decipad/editor';

<Meta
  title="Legacy/Editor/Blocks/CodeBlock"
  parameters={{
    docs: {
      source: {
        code: `<CodeBlock
  element={{
    children: [{ text: "a = 10 apples b = 10 apples a + b" }],
  }}
  attributes={{ 'data-slate-node': 'element', ref: null }}
>
  a = 10 apples b = 10 apples a + b
</CodeBlock>`,
      },
    },
  }}
  argTypes={{
    children: {
      defaultValue: `a = 10 apples
b = 10 apples
a + b`,
      description: 'The content of the code block',
      type: 'string',
    },
  }}
  component={CodeBlock}
/>

# CodeBlock

The CodeBlock component is one of the most important components for our editor, this is where the user will do all of their calculations and decisions. The CodeBlock component gets automatic syntax highlighting once it is plugged into slate. The component itself is a slate specific component where it requires a result to be either undefined or a string to output the calculation's result.

export const Template = (args) => (
  <ResultsContextProvider
    value={{test: { type: Type.Number, value: [20], errors: [] }}}
  >
    <CodeBlock
      element={{ id: 'test' }}
      {...args}
    />
  </ResultsContextProvider />
);

export const NoResult = (args) => (
  <ResultsContextProvider value={{test: null}}>
    <CodeBlock element={{ id: 'test', children: [] }} {...args} />
  </ResultsContextProvider />
);

<Canvas>
  <Story name="With result">{Template.bind({})}</Story>
  <Story name="Without result">{NoResult.bind({})}</Story>
</Canvas>

<Props story="With result" />
<Props story="Without result" />
