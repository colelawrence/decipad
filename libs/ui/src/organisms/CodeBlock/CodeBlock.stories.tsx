import { Meta, Story } from '@storybook/react';
import { SerializedType } from '@decipad/language';
import { docs } from '@decipad/routing';
import { withResults } from '../../storybook-utils/results';
import { CodeLine } from '..';
import { CodeBlock } from './CodeBlock';

const block = {
  blockId: 'block-one',
  isSyntaxError: false,
  results: [
    {
      blockId: 'block-one',
      statementIndex: 0,
      value: ['Lorem', 'Ipsum', 'Dolor', 'Sit', 'Amet'],
      type: {
        kind: 'column',
        cellType: { kind: 'string' },
        columnSize: 5,
        indexedBy: null,
      } as SerializedType,
      visibleVariables: new Set() as ReadonlySet<string>,
    },
  ],
};

const lines = [
  {
    displayInline: true,
    syntaxError: {
      message: 'SyntaxError',
      url: docs({}).$,
    },
  },
  {
    displayInline: true,
    result: {
      type: block.results[0].type,
      value: block.results[0].value,
    },
  },
];

export default {
  title: 'Organisms / Editor / Code / Block',
  component: CodeBlock,
  decorators: [
    withResults({
      [block.blockId]: block,
    }),
  ],
} as Meta;

export const Normal: Story = () => (
  <CodeBlock blockId={block.blockId} expandedResult={block.results[0]}>
    <CodeLine {...lines[0]}>42 + 1337;</CodeLine>
    <CodeLine {...lines[1]}>
      ["Lorem", "Ipsum", "Dolor", "Sit", "Amet"]
    </CodeLine>
  </CodeBlock>
);
