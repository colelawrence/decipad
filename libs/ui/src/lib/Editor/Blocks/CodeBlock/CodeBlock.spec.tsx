import React from 'react';
import { render } from 'test-utils';
import { CodeBlock } from './CodeBlock.component';

describe('CodeBlock Block', () => {
  it('renders', () => {
    const { container } = render(
      <CodeBlock
        element={{
          children: [{ text: 'a = 10 apples \n b = 10 apples \n a + b' }],
          result: '20 apples',
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        a = 10 apples b = 10 apples a + b
      </CodeBlock>
    );

    expect(container).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <CodeBlock
        element={{
          children: [{ text: 'a = 10 apples \n b = 10 apples \n a + b' }],
          result: '20 apples',
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        a = 10 apples b = 10 apples a + b
      </CodeBlock>
    );

    expect(container).toMatchSnapshot();
  });
});
