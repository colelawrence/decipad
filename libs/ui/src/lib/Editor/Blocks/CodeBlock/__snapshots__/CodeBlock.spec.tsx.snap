// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock Block matches snapshot 1`] = `
.emotion-1 {
  display: inline-block;
  font-family: var(--chakra-fonts-mono);
  font-size: var(--chakra-fontSizes-sm);
  -webkit-padding-start: 0.2em;
  padding-inline-start: 0.2em;
  -webkit-padding-end: 0.2em;
  padding-inline-end: 0.2em;
  border-radius: var(--chakra-radii-sm);
  background: var(--chakra-colors-gray-100);
  color: var(--chakra-colors-gray-800);
  width: 100%;
  padding: var(--chakra-space-5);
  border-radius: 2px;
  background: var(--chakra-colors-gray-100);
}

.emotion-2 {
  background: rgb(252,252,263);
  color: var(--chakra-colors-black);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
  -webkit-padding-start: var(--chakra-space-6);
  padding-inline-start: var(--chakra-space-6);
  -webkit-padding-end: var(--chakra-space-6);
  padding-inline-end: var(--chakra-space-6);
  display: inline-block;
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: var(--chakra-space-3);
  width: 100%;
}

<div>
  <div
    class="emotion-0"
  >
    <code
      class="chakra-code emotion-1"
      data-slate-node="element"
    >
      a = 10 apples b = 10 apples a + b
    </code>
    <div
      class="emotion-2"
      contenteditable="false"
    >
      20
       &lt;number&gt;
    </div>
  </div>
</div>
`;
